File: DBMS_Lab_Project/StorageManager.cpp
#include "StorageManager.h"
#include <fstream>
#include <sstream>
#include <stdexcept>
#include "BinaryUtils.h"
StorageManager::StorageManager(const std::string& path) : filepath(path) {}
void StorageManager::load(DBMS& dbms) {
    std::ifstream file(filepath, std::ios::binary);
    if (!file.is_open()) return;
    dbms.clear();
    
    // Check if file is empty
    file.peek();
    if(file.eof()) return;

    size_t total = 0;
    readSize(file, total);

    for (size_t i = 0; i < total; ++i) {
        if (file.peek() == EOF) break;
        std::string type, name, data;
        readString(file, type);
        readString(file, name);
        readString(file, data);
        dbms.loadStructure(type, name, data);
    }
}
void StorageManager::save(const DBMS& dbms) {
    std::ofstream file(filepath, std::ios::trunc | std::ios::binary);
    if (!file.is_open()) throw std::runtime_error("Could not open file for writing: " + filepath);
    std::string content = dbms.serializeAll();
    file.write(content.data(), content.size());
}

File: DBMS_Lab_Project_Go/internal/dbms/storage_manager.go
package dbms

import (
	"dbms_lab_project/internal/datastructures"
	"io"
	"os"
)

type StorageManager struct {
	filepath string
}

func NewStorageManager(path string) *StorageManager {
	return &StorageManager{filepath: path}
}

func (sm *StorageManager) Load(db *DBMS) error {
	file, err := os.Open(sm.filepath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil // File doesn't exist, start with empty DB
		}
		return err
	}
	defer file.Close()

	db.Clear()
	total, err := datastructures.ReadSize(file)
	if err != nil {
		if err == io.EOF {
			return nil
		}
		return err
	}

	for i := 0; i < total; i++ {
		typeStr, err := datastructures.ReadString(file)
		if err != nil { return err }
		name, err := datastructures.ReadString(file)
		if err != nil { return err }
		data, err := datastructures.ReadString(file)
		if err != nil { return err }
		db.LoadStructure(typeStr, name, data)
	}
	return nil
}

func (sm *StorageManager) Save(db *DBMS) error {
	file, err := os.Create(sm.filepath)
	if err != nil {
		return err
	}
	defer file.Close()

	_, err = file.WriteString(db.SerializeAll())
	return err
}
	if err != nil {
		return err
	}
	defer file.Close()

	content := db.SerializeAll()
	_, err = file.WriteString(content)
	return err
}
